File: Go/350.go
/*
Create a map to track count of each integer appeared in nums1,
then go thru nums2 and put into intersection if the char has count > 0 in the map,
decrement the count everytime it's put into intersection.
*/
func intersect(nums1 []int, nums2 []int) []int {
    
    dict, res := make(map[int]int), make([]int, 0, len(nums2))
    for _, n := range nums1 {
        dict[n] += 1
    }
    for _, n := range nums2 {
        if dict[n] > 0 {
            res = append(res, n)
            dict[n] -= 1
        }
    }
    return res[:]
}

File: Java/350.java
class Solution {
    public int[] intersect(int[] nums1, int[] nums2) {
        if(nums1 == null || nums2 == null || nums1.length == 0 || nums2.length == 0) return new int[]{};
        Map<Integer, Integer> counts = new HashMap<>();
        List<Integer> resList = new ArrayList<>();
        
        for(int n : nums1) {
            counts.put(n, counts.getOrDefault(n, 0) + 1);
        }
        
        for(int n : nums2) {
            int count = counts.getOrDefault(n, 0);
            if(count > 0) {
                resList.add(n);
                counts.put(n, count-1);
            }
        }
        
        int[] res = new int[resList.size()];
        int i=0;
        for(int n : resList) {
            res[i++] = n;
        }
        return res;
    }
}
